### PR TITLE
Common metric extraction [pLDDT, MSA, PAE]: extract_ouput.py -> extract_metrics.py in every program module

### PR DESCRIPTION
**Pull request by keiran-rowell-unsw**
Moving towards having a single point of metric extraction that uses `BioPython` for structures etc. Extracting from fixed-width .pdb files causes issues (#226, #251).

The idea is that all programs actually produce the same set of metrics:
-  PDB/mmCIF structures (**pdb**)
- Per-residue confidence score (**plddt**)
- Multiple Sequence Alignment (**msa**)
- Inter-residue positioning Pair Alignment Error (**pae**)

These are stored in different ways be each program, which I've collated in other Issues MSA: (#265), PAE: (#262).

The idea is `extract_metrics.py` sits immediately after the AI protein structure prediction in `/modules/local/run_[program]/main.nf` and dumps all these metrics (if available) as `${meta.id}_[metric].tsv` in the directory of execution.  If some metric _may_ not be available it is an `optional: true` emit:. Happy to have another opinionated approach

This means `generate_report.py` can be largely indifferent to source program, giving consistent pipelines _etc_.

**Not included in current PR** 
- Using the newly emitted metrics in downstream pipelines. Workflow comp ability will be solve in a future
- ColabFold to use `extract_metrics.py` -- Currently it passes on ColabFold's figures, but I have an implementation of raw metrics in my dev branch
- Boltz to use `extract_metrics.py` -- This is fresh inclusion and the `.npz` files are passed through, rather than having `emit: [metric]` pass through raw data 

**Future benefits**
- PAE data is now available
- Boltz includes PAE files  with [--write_full_pae --output_format pdb](https://github.com/jwohlwend/boltz/blob/main/docs/prediction.md) options
- Report can be altered to include PAE data necessary to assess scientific quality (#303)
- A more general and simple MultiQC/proteinfold module can create a 'bulk' report from globbing all the raw `[metric].tsv` files

**Future renames**
- Since currently only plddt is used for mqc, I've populated `emit: multiqc` with the `${meta.id}_plddt.tsv`
- I'd like to call it `emit: plddt` and have a `multiqc` values downstream
- Simplify `ch_report_input` to pass raw data into `generate_report.py`

## PR checklist

- [X] This comment contains a description of changes (with reason).
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.

### CODE REVIEW

**Review by JoseEspinosa — COMMENTED** [review]
This is a very nice addition! However, I was wondering whether we should create a separate process to run `extract_metrics.py`. This is something I've been thinking about for a while. The reasoning behind this suggestion is that, if there's an error when running the Python script, the output from prediction tools (which is much more computationally intensive) won't be lost. Do you think it's worth the hassle?
If yes, maybe we can create an issue and do it in a different PR

**Comment by keiran-rowell-unsw** [comment]
> This is a very nice addition! However, I was wondering whether we should create a separate process to run `extract_metrics.py`. This is something I've been thinking about for a while. The reasoning behind this suggestion is that, if there's an error when running the Python script, the output from prediction tools (which is much more computationally intensive) won't be lost. Do you think it's worth the hassle? If yes, maybe we can create an issue and do it in a different PR

Yes I agree this is a better approach, @ziadbkh also had the same thought. I figured it was easier to have a visible Python script that can be wrapped in a process.

If making a new process, these need to be in the process container:
- import numpy as np
- #import torch moved to a conditional import [if .pt exists] since too bulky [to] import if not used

It was nice working with `generate_report.py`s environment in its own module.

File: modules/local/run_alphafold2_pred/main.nf [hunk 1]
```diff
@@ -21,15 +21,20 @@ process RUN_ALPHAFOLD2_PRED {
...
     path ("${fasta.baseName}*")
-    tuple val(meta), path ("${meta.id}_alphafold2.pdb")   , emit: top_ranked_pdb
-    tuple val(meta), path ("${fasta.baseName}/ranked*pdb"), emit: pdb
-    tuple val(meta), path ("*_msa.tsv")                   , emit: msa
-    tuple val(meta), path ("*_mqc.tsv")                   , emit: multiqc
-    path "versions.yml"                                   , emit: versions
+    tuple val(meta), path ("${meta.id}_alphafold2.pdb")     , emit: top_ranked_pdb
+    tuple val(meta), path ("${fasta.baseName}/ranked*.pdb") , emit: pdb
+    tuple val(meta), path ("${meta.id}_msa.tsv")            , emit: msa
+    // TODO: re-label multiqc -> plddt so multiqc channel can take in all metrics
```
Comment (JoseEspinosa) [inline modules/local/run_alphafold2_pred/main.nf:31]:
```suggestion
    // TODO: re-label multiqc -> plddt so multiqc channel can take in all metrics
```

File: modules/local/run_helixfold3/main.nf [hunk 2]
```diff
@@ -64,20 +66,14 @@ process RUN_HELIXFOLD3 {
...
-    done
-    cd ..
+    cp "${fasta.baseName}/${fasta.baseName}-rank1/predicted_structure.pdb" "./${meta.id}_helixfold3.pdb"
+    cp "${fasta.baseName}/${fasta.baseName}-rank1/predicted_structure.cif" "./${meta.id}_helixfold3.cif"
+   
+    
+    extract_output.py --name ${meta.id} \\
+        --structs "${fasta.baseName}/${fasta.baseName}-rank*/predicted_structure.pdb" \\
+        --pkls "${fasta.baseName}/final_features.pkl" \\
+        --jsons "${fasta.baseName}-rank*/all_results.json"
```
Comment (JoseEspinosa) [inline modules/local/run_helixfold3/main.nf:76]:
```suggestion
        --jsons "${fasta.baseName}-rank*/all_results.json"
```

File: modules/local/run_rosettafold_all_atom/main.nf [hunk 3]
```diff
@@ -17,30 +17,33 @@ process RUN_ROSETTAFOLD_ALL_ATOM {
...
     $args
 
-    cp "${yaml.baseName}.pdb" "${meta.id}_rosettafold_all_atom.pdb"
-    awk '{printf "%s\\t%.0f\\n", \$6, \$11 * 100}' ${meta.id}_rosettafold_all_atom.pdb | uniq > plddt.tsv
-    echo -e Positions"\\t"${meta.id}_rosettafold_all_atom.pdb > header.tsv
-    cat header.tsv plddt.tsv > "${meta.id}_plddt_mqc.tsv"
+    mamba run --name RFAA extract_metrics.py --name ${meta.id} \\
+      --structs "${yaml.baseName}_rosettafold_all_atom.pdb" \\
+      --a3ms "${yaml.baseName}/A/t000_.msa0.a3m" \\
+      --pts ${yaml.baseName}_aux.pt
```
Comment (JoseEspinosa) [inline modules/local/run_rosettafold_all_atom/main.nf:46]:
```suggestion
        --structs "${yaml.baseName}_rosettafold_all_atom.pdb" \\
        --a3ms "${yaml.baseName}/A/t000_.msa0.a3m" \\
        --pts ${yaml.baseName}_aux.pt
```

**Review by JoseEspinosa — APPROVED** [review]
OK, I think it will better to implement the new process on a new PR. Now, I will try to make lint happy with my suggestions and then merge this one.
Really nice addition to the pipeline ❤️

**Review by JoseEspinosa — CHANGES_REQUESTED** [review]
Seems that the option `Allow edits from maintainers` is not enable on this PR and I can not commit my suggestions, please do it and we can merge it once lint is happy.

**Comment by JoseEspinosa** [comment]
> > This is a very nice addition! However, I was wondering whether we should create a separate process to run `extract_metrics.py`. This is something I've been thinking about for a while. The reasoning behind this suggestion is that, if there's an error when running the Python script, the output from prediction tools (which is much more computationally intensive) won't be lost. Do you think it's worth the hassle? If yes, maybe we can create an issue and do it in a different PR
> 
> Yes I agree this is a better approach, @ziadbkh also had the same thought. I figured it was easier to have a visible Python script that we may then choose to wrap in a process.
> 
> If making a new process, these need to be in the process container:
> 
>     * `import numpy as np`
> 
>     * `#import torch` - moved to a conditional import [if .pt exists] since too bulky [to] import if not used
> 
> 
> It was nice working with `generate_report.py`s environment in its own module.

Opened issue --> https://github.com/nf-core/proteinfold/issues/309

**Review by JoseEspinosa — APPROVED** [review]
LGTM Thanks! 🚀
